contrib/win32/win32compat/misc.c

```diff
@@ -2119,3 +2124,97 @@ strrstr(const char *inStr, const char *pattern)
 
 	return last;
 }
+
+int
+add_mark_of_web(const char* filename)
+{
+	if (motw_zone_id < 5) {
+		char* fileStreamPath = NULL;
+		size_t fileStreamPathLen = strlen(filename) + strlen(":Zone.Identifier") + 1;
+
+		fileStreamPath = malloc(fileStreamPathLen * sizeof(char));
+
+		if (fileStreamPath == NULL) {
+			return -1;
+		}
+
+		sprintf_s(fileStreamPath, fileStreamPathLen, "%s:Zone.Identifier", filename);
+
+		char* zoneIdentifierStr = NULL;
+		size_t zoneIndentifierLen = strlen("[ZoneTransfer]\nZoneId=") + 1 + 1;
+
+		zoneIdentifierStr = malloc(zoneIndentifierLen * sizeof(char));
+
+		if (zoneIdentifierStr == NULL) {
+			return -1;
+		}
+
+		sprintf_s(zoneIdentifierStr, zoneIndentifierLen, "[ZoneTransfer]\nZoneId=%d", motw_zone_id);
+
+		int ofd, status = 0;
+
+		// create zone identifer file stream and then write the Mark of the Web to it
+		if ((ofd = open(fileStreamPath, O_WRONLY | O_CREAT, USHRT_MAX)) == -1) {
+			status = -1;
+			goto cleanup;
+		}
+
+		if (atomicio(vwrite, ofd, zoneIdentifierStr, zoneIndentifierLen) != zoneIndentifierLen) {
+			status = -1;
+		}
+
+		if (close(ofd) == -1) {
+			status = -1;
+		}
+
+	cleanup:
+		free(fileStreamPath);
+		free(zoneIdentifierStr);
+		return status;
+	}
+	else {
+		return -1;
+	}
+}
+
+/* Gets the zone identifier value based on the provided hostname, 
+and sets the global motw_zone_id variable with that value. */
+void get_zone_identifier(const char* hostname) {
+	static const CLSID CLSID_ISM =
+	{ 0x7B8A2D94, 0x0AC9, 0x11D1,
+	{ 0x89, 0x6C, 0x00, 0xC0, 0x4F, 0xB6, 0xBF, 0xC4 } };
+	static const IID IID_IISM =
+	{ 0x79EAC9EE, 0xBAF9, 0x11CE,
+	{ 0x8C, 0x82, 0x00, 0xAA, 0x00, 0x4B, 0xA9, 0x0B } };
+	IInternetSecurityManager* IISM;
+	CoInitialize(NULL);
+	HRESULT hr = CoCreateInstance(&CLSID_ISM, NULL,
+		CLSCTX_ALL, &IID_IISM, (void**)&IISM);
+	if (SUCCEEDED(hr) || hr == RPC_E_CHANGED_MODE)
```

Comment:
We could do early return here on fail.  Also, I think we should always set `motw_zone_id = 5` on the fail case, even though we initialize it to 5.  We don't want to inadvertently pick up a stale setting.